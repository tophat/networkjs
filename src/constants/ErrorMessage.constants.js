import NetworkStatuses from './NetworkStatus.constants'

export const ErrorMessage = {
    INVALID_EVENT: `Event must be one of ${NetworkStatuses}`,
}

export const ErrorMessages = Object.values(ErrorMessage)
