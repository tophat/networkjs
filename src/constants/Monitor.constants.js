export const Monitor = {
    NETWORK: 'network',
    SERVICE: 'service',
    STABILITY: 'stability',
}

export const Monitors = Object.values(Monitor)
