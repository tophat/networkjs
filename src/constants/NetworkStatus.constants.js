export const NetworkStatus = {
    ONLINE: 'online',
    OFFLINE: 'offline',
    STABLE: 'stable',
    UNSTABLE: 'unstable',
    DEGRADED: 'degraded',
    RESOLVED: 'resolved',
}

export const NetworkStatuses = Object.values(NetworkStatus)
