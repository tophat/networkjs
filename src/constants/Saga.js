export const Saga = {
    NETWORK: 'network',
    SERVICE: 'service',
    STABILITY: 'stability',
}

export const Sagas = Object.values(Saga)
