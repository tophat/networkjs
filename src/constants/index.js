export { NetworkStatus, NetworkStatuses } from './NetworkStatus.constants'
export { Monitor, Monitors } from './Monitor.constants'
export { ErrorMessage, ErrorMessages } from './ErrorMessage.constants'
