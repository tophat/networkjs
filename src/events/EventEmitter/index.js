import Event from '../Event'

export default class EventEmitter {
    constructor() {
        this.events = {}
    }

    registerEvent(eventName) {
        const event = new Event(eventName)
        this.events[eventName] = event
    }

    dispatchEvent(eventName, ...eventArgs) {
        this.events[eventName].callbacks.forEach(callback => {
            callback(...eventArgs)
        })
    }

    addEventListener(eventName, callback) {
        this.events[eventName].registerCallback(callback)
    }
}
