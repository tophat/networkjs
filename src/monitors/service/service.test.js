import EventEmitter from '../../events/EventEmitter'
import { NetworkStatus } from '../../constants'
import ServiceMonitor from '.'

jest.mock('../../events/EventEmitter')

const FAILURE_THRESHOLD = 2
const DECREMENT_TIME = 5000

describe('Service Monitor', () => {
    let monitor
    const emitter = new EventEmitter()

    beforeEach(() => {
        jest.useFakeTimers()

        monitor = new ServiceMonitor(emitter, {
            failureThreshold: FAILURE_THRESHOLD,
            decrementTime: DECREMENT_TIME,
        })
    })

    describe('constructor', () => {
        it('initializes with the correct props', () => {
            expect(monitor).toMatchSnapshot()
        })
    })

    describe('initialize', () => {
        it('initializes failure counts', () => {
            monitor.initialize()

            expect(monitor.paused).toBe(false)
            expect(monitor.failureCounts).toStrictEqual({ '*': 0 })
        })
    })

    describe('pause', () => {
        it('sets paused to true', () => {
            monitor.pause()
            expect(monitor.paused).toBe(true)
        })
    })

    describe('resume', () => {
        it('re-initializes monitor', () => {
            monitor.pause()
            expect(monitor.paused).toBe(true)

            monitor.resume()
            expect(monitor.paused).toBe(false)
        })
    })

    describe('error handler', () => {
        it('does nothing if paused', () => {
            monitor.pause()
            monitor.handleError('/path/1', 502)

            expect(monitor.failureCounts['*']).toBe(0)
            expect(setTimeout).not.toHaveBeenCalled()
        })

        it('does nothing if status not tracked', () => {
            monitor.handleError('/path/1', 400)

            expect(monitor.failureCounts['*']).toBe(0)
            expect(setTimeout).not.toHaveBeenCalled()
        })

        it('increments failure count and calls waitAndDecrement', () => {
            jest.spyOn(monitor, 'waitAndDecrement')
            monitor.handleError('/path/1', 502)

            expect(monitor.failureCounts['*']).toBe(1)
            expect(monitor.emitter.dispatchEvent).not.toHaveBeenCalled()
            expect(monitor.waitAndDecrement).toHaveBeenCalledTimes(1)
        })

        it('emits service degraded event when threshold is hit', () => {
            jest.spyOn(monitor, 'waitAndDecrement')
            monitor.handleError('/path/1', 502)
            monitor.handleError('/path/1', 502)

            expect(monitor.failureCounts['*']).toBe(2)
            expect(monitor.emitter.dispatchEvent).toHaveBeenCalledWith(
                NetworkStatus.DEGRADED,
                '*',
            )
            expect(monitor.waitAndDecrement).toHaveBeenCalledTimes(2)
        })

        it('does nothing if not tracking path', () => {
            monitor = new ServiceMonitor(emitter, {
                definitions: [
                    {
                        name: 'path 1',
                        regex: new RegExp('/path/1'),
                    },
                ],
            })

            jest.spyOn(monitor, 'waitAndDecrement')
            monitor.handleError('/path/2', 502)

            expect(monitor.failureCounts['path 1']).toBe(0)
            expect(monitor.waitAndDecrement).toHaveBeenCalledTimes(0)
        })
    })

    describe('wait and decrement', () => {
        beforeEach(() => {
            jest.restoreAllMocks()
        })

        it('calls set setTimeout with the provided decrement time', () => {
            monitor.waitAndDecrement('*')

            expect(setTimeout).toHaveBeenCalledWith(
                expect.any(Function),
                DECREMENT_TIME,
            )
        })

        it('does nothing if paused', () => {
            const decrementSpy = jest.spyOn(monitor, 'decrement')

            monitor.pause()
            monitor.waitAndDecrement('*')

            jest.runAllTimers()
            expect(decrementSpy).not.toHaveBeenCalled()
        })

        it('decrements failure count', () => {
            monitor.failureCounts = { '*': 1 }
            monitor.waitAndDecrement('*')

            jest.runAllTimers()
            expect(monitor.failureCounts['*']).toBe(0)
        })

        it('emits service resolved event after falling under threshold', () => {
            monitor.failureCounts = { '*': 2 }
            monitor.waitAndDecrement('*')

            jest.runAllTimers()

            expect(monitor.failureCounts['*']).toBe(1)
            expect(monitor.emitter.dispatchEvent).toHaveBeenCalledWith(
                NetworkStatus.RESOLVED,
                '*',
            )
        })
    })
})
