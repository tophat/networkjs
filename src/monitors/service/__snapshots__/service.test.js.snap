// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Monitor constructor initializes with the correct props 1`] = `
ServiceMonitor {
  "decrementTime": 5000,
  "definitions": Array [
    Object {
      "name": "*",
      "regex": /\\.\\*/,
    },
  ],
  "emitter": EventEmitter {
    "addEventListener": [MockFunction],
    "dispatchEvent": [MockFunction],
    "registerEvent": [MockFunction],
  },
  "failureCounts": Object {
    "*": 0,
  },
  "failureThreshold": 2,
  "paused": false,
  "statuses": Array [
    502,
    503,
    504,
  ],
}
`;
